File: Java/Heranca/exercicio3.java
interface Animal {
    String fazerSom();
    String getNome();
}

interface AnimalDomestico {
    String brincar();
    String alimentar();
}

class Cachorro implements Animal, AnimalDomestico {
    private String nome;

    public Cachorro(String nome) {
        this.nome = nome;
    }

    @Override
    public String fazerSom() {
        return "Au Au";
    }

    @Override
    public String getNome() {
        return nome;
    }

    @Override
    public String brincar() {
        return "Brincando";
    }

    @Override
    public String alimentar() {
        return "Alimentando";
    }
}

class Gato implements Animal, AnimalDomestico {
    private String nome;

    public Gato(String nome) {
        this.nome = nome;
    }

    @Override
    public String fazerSom() {
        return "Miau";
    }

    @Override
    public String getNome() {
        return nome;
    }

    @Override
    public String brincar() {
        return "Brincando";
    }

    @Override
    public String alimentar() {
        return "Alimentando";
    }
}

public class exercicio3 {
    public static void main(String[] args) {
        // Teste
        Cachorro cachorro = new Cachorro("Flokus");
        Gato gato = new Gato("Felicete");

        System.out.println("Cachorro:");
        System.out.println("Nome: " + cachorro.getNome());
        System.out.println("Faz som: " + cachorro.fazerSom());
        System.out.println("Brincando: " + cachorro.brincar());

        System.out.println("\nGato:");
        System.out.println("Nome: " + gato.getNome());
        System.out.println("Faz som: " + gato.fazerSom());
        System.out.println("Alimentando: " + gato.alimentar());
    }
}

File: PrimeiraParte/C++/Heranca/exercicio3.cpp
#include <iostream>
#include <string>

class Animal {
public:
    Animal(std::string nome) : nome(nome) {}

    virtual std::string fazerSom() {
        return "Som genérico do animal";
    }

    std::string getNome() {
        return nome;
    }

private:
    std::string nome;
};

class AnimalDomestico {
public:
    std::string brincar() {
        return "Brincando";
    }

    std::string alimentar() {
        return "Alimentando";
    }
};

class Cachorro : public Animal, public AnimalDomestico {
public:
    Cachorro(std::string nome, std::string cor) : Animal(nome), cor(cor) {}

    std::string fazerSom() override {
        return "Au Au";
    }

private:
    std::string cor;
};

class Gato : public Animal, public AnimalDomestico {
public:
    Gato(std::string nome, std::string cor) : Animal(nome), cor(cor) {}

    std::string fazerSom() override {
        return "Miau";
    }

private:
    std::string cor;
};

int main() {
    // Teste
    Cachorro cachorro("Flokus", "Marrom");
    Gato gato("Felicete", "Branco");

    std::cout << "Cachorro:\n";
    std::cout << "Nome: " << cachorro.getNome() << "\n";
    std::cout << "Faz som: " << cachorro.fazerSom() << "\n";
    std::cout << "Brincando: " << cachorro.brincar() << "\n";

    std::cout << "\nGato:\n";
    std::cout << "Nome: " << gato.getNome() << "\n";
    std::cout << "Faz som: " << gato.fazerSom() << "\n";
    std::cout << "Alimentando: " << gato.alimentar() << "\n";

    return 0;
}

